base/applications/utilman/registry.c

```diff
@@ -0,0 +1,201 @@
+                      OUT PVOID ReturnedData)
+{
+    LONG lResult;
+    HKEY hKeyQueryValue;
+    DWORD cbSize = sizeof(ReturnedData);
+
+    lResult = RegOpenKeyExW(hKey,
+                            lpwszSubKey,
+                            0,
+                            KEY_READ,
+                            &hKeyQueryValue);
+    if (lResult != ERROR_SUCCESS)
+    {
+        DPRINT("QueryAppSettings(): Failed to open the key of path \"%S\". The error code is \"%li\".\n", lpwszSubKey, lResult);
+        return FALSE;
+    }
+
+    lResult = RegQueryValueExW(hKeyQueryValue,
+                               lpwszRegValue,
+                               0,
+                               0,
+                               (LPBYTE)&ReturnedData,
+                               &cbSize);
+    if (lResult != ERROR_SUCCESS)
+    {
+        DPRINT("QueryAppSettings(): Failed to query the data from value \"%S\". The error code is \"%li\".\n", lpwszRegValue, lResult);
+        RegCloseKey(hKeyQueryValue);
+        return FALSE;
+    }
+
+    if (cbSize != sizeof(ReturnedData))
+    {
+        DPRINT("QueryAppSettings(): The buffer's size is too small to hold the actual buffer's content! Bail out!\n");
+        RegCloseKey(hKeyQueryValue);
+        return FALSE;
+    }
+
+    RegCloseKey(hKeyQueryValue);
+    return TRUE;
+}
+
+/**
+ * @SaveAppSettings
+ *
+ * Save an application's setting data to the Registry. This is a function helper.
+ *
+ * @param[in]   hKey
+ *     A handle to a key.
+ *
+ * @param[in]   lpwszRegValue
+ *      The path to the sub key where the value needs to be created.
+ *
+ * @param[out]   dwRegType
+ *      The type of registry value to be created (e.g. a REG_DWORD).
+ *
+ * @param[out]   Data
+ *      A pointer to an arbitrary data for the value to be set. Being arbitrary,
+ *      the data can be of any type (in conformity with the registry type pointed by
+ *      dwRegType) otherwise the function might lead to a undefined behaviour.
+ *
+ * @return
+ *      Returns TRUE if the function successfully saved the application's setting,
+ *      FALSE otherwise for failure.
+ *
+ */
+BOOL SaveAppSettings(IN HKEY hKey,
+                     IN LPCWSTR lpwszRegValue,
+                     IN DWORD dwRegType,
+                     IN PVOID Data)
+{
+    LONG lResult;
+    HKEY hKeySetValue;
+
+    lResult = RegOpenKeyExW(hKey,
+                            NULL,
+                            0,
+                            KEY_SET_VALUE,
+                            &hKeySetValue);
+    if (lResult != ERROR_SUCCESS)
+    {
+        DPRINT("SaveAppSettings(): Failed to open the key, the error code is \"%li\"!\n", lResult);
+        return FALSE;
+    }
+
+    lResult = RegSetValueExW(hKeySetValue,
+                             lpwszRegValue,
+                             0,
+                             dwRegType,
+                             (LPBYTE)&Data,
+                             sizeof(Data));
```

Comment:
Nope: sizeof(Data) == size of pointer in the current architecture (== 4 bytes in 32-bit, 8 bytes in 64-bit). You should specify instead a variable (passed via a separate parameter in SaveAppSettings) that contains the actual size of the data buffer you want to save.